docs/rfc/active/tags/README.md

```diff
@@ -0,0 +1,210 @@
+
+- Ability to associate tags with any type of entity, even other tags!
+- Ability to tag the same entity with multiple tags.
+- Ability to tag multiple objects with the same tag instance.
+- To the point above, ability to make easy tag-based searches later on.
+- Metadata on tags is TBD
+
+### Extensibility
+
+The normal new-entity-onboarding work is obviously required.
+
+Hopefully this can serve as a stepping stone to work on special cases such as the tag-based privacy tagging mentioned in
+the roadmap.
+
+## Non-Requirements
+
+Let's leave the UI work required for this to another time.
+
+## Detailed design
+
+We want to introduce some new under `datahub/metadata-models/src/main/pegasus/com/linkedin/common/`.
+
+### `Tag` entity
+
+First we create a `TagMetadata` entity, which defines the actual tag-object.
+
+The edit property defines the edit rights of the tag, as some tags (like sensitivity tags) should be read-only for a
+majority of users
+
+```
+/**
+ * Tag information
+ */
+record TagMetadata {
+   /**
+   * Tag URN, e.g. urn:li:tag:<name>
+   */
+   urn: TagUrn
+
+   /**
+   * Tag value.
+   */
+   value: string
+
+   /**
+   * Optional tag description
+   */
+   description: optional string
+
+   /**
+   * Optional reference to an external definition
+   */
+   externalDefinition: optional ExternalDefinition
+
+   /**
+   * Audit stamp associated with creation of this tag
+   */
+   createStamp: AuditStamp
+}
+```
+
+here ExternalDefinition is simply
+
+```
+/**
+ * External definition for a Tag
+ */
+record ExternalDefinition {
+  /**
+   * Source of the definition
+   */
+  source: optional string
+  /**
+   * URL reference for the external definition
+   */
+  reference: optional string
+}
+```
+
+### `TagAttachment`
+
+We define a `TagAttachment`-model, which describes the application of a tag to a entity
+
+```
+/**
+ * Tag information
+ */
+record TagAttachment {
+
+  /**
+   * Tag in question
+   */
+  tag: TagUrn
+
+  /**
+   * Who has edit rights to this employment.
+   * WIP, pending access-control support in Datahub.
+   * Relevant for privacy tags at least.
+   * We might also want to add view rights?
+   */
+  edit: union[None, any, role-urn]
+
+   /**
+   * Audit stamp associated with employment of this tag to this entity
+   */
+   applicationStamp: AuditStamp
+}
+```
+
+### `Tags` container entity
+
+Then we define a `Tags`-model, which is used as a container for tag employments. This container is taken into use in
+aspects.
+
+```
+namespace com.linkedin.common
+
+/**
+ * Tags information
+ */
+record Tags {
+
+   /**
+   * List of tag employments
+   */
+   elements: array[TagAttachment] = [ ]
+}
+```
+
+## How we teach this
+
+We should create/update user guides to educate users for:
+
+- Suggestions on how to use tags: low threshold metadata-addition, and the possibility of doing new types of searches
+
+## Drawbacks
+
+This is definitely more complex than just adding strings to an array.
+
+## Alternatives
```

Comment:
I know it sounds a little bit weird to model and implement a tag system for metadata in a non-metadata model way. Fundamentally, a minimal tag solution requirements are: 1) link a string to an ID of something; 2) provide ID to tags and tag to IDs search. Have you considered to design and impl this tag solution as a built-in feature for a normal application? Will it be a even flexible design?

Reply:
Sure, the ID-> string approach is one way, but you still need to "attach" it to objects, and I guess you'd want to have the ability to tag a object with more than one tag, so some sort of a "container" is needed. This is basically what we have right now. The only extra properties we're introducing are some AuditStamps and a descriptions (@jplaisted  was worried that tags go undocumented). The permission-to-edit functionality is an open question, and can be removed if it adds too much complexity. But I can mention this view in this section. Thanks!

>  built-in feature for a normal application? Will it be a even flexible design?

Sorry, I'm not sure what you mean by this? Can you elaborate, please?

Reply:
please ignore my below comment, it's implementation details.

>built-in feature for a normal application? Will it be a even flexible design?